include/zephyr/drivers/cellular.h

```diff
@@ -210,6 +229,33 @@ static inline int cellular_get_modem_info(const struct device *dev,
 	return api->get_modem_info(dev, type, info, size);
 }
 
+/**
+ * @brief Get network registration status for the device
+ *
+ * @param dev Cellular network device instance
+ * @param tech Which access technology to get status for
+ * @param status Registration status for given access technology
+ * @param reject_cause 3GPP TS 24.301 Annex A reject cause code, if status indicates denied
```

Comment:
This value we need to enumerate and document like the status and access tech, we may also want to bundle it with the status to make the API a bit cleaner, something like
```
struct cellular_registration {
        enum cellular_registration_status *status;
        uint8_t *reject_cause;
};
```

Reply:
Or even better, we can make the status data a union

enum cellular_registration_status *status;

struct cellular_registration_status_rejected {
        uint8_t reject_cause;
};

union cellular_registration_status_data {
        struct cellular_registration_status_rejected rejected;
};

struct cellular_registration {
        enum cellular_registration_status *status;
        union cellular_registration_status_data status_data;
};

It may be overkill, what do you think?

Reply:
> This value we need to enumerate and document like the status and access tech, we may also want to bundle it with the status to make the API a bit cleaner, something like...

I'm not following you here, do you mean we should create an enum for every single reject cause as showed in https://www.etsi.org/deliver/etsi_ts/124300_124399/124301/15.04.00_60/ts_124301v150400p.pdf (Look at Annex A)? That would be one huge enum

In addition, I do get the bundling of registration status and reject cause like this, I'm buying that.
```
struct cellular_registration {
        enum cellular_registration_status *status;
        uint8_t *reject_cause;
};
```
However I do disagree on two things:
1. It looks like `status` is a pointer to a living variable inside the driver. I think the returned value should be value at the moment of the call, since if the value changes let's say in the middle of an if-statement in the caller, sounds dangerous.
2. In my opinion, structs that require accesses like this `registration->data.denied.reject_cause` quite severely damages the overview and readability of the code, since it's more layers, requiring you to go back and forth to the struct, and is it a pointer or variable, and so on. Sure, helpful in some cases but I think in this case it's overkill, if you ask me :slightly_smiling_face: